File: src/lib/nodes/competitive.ts
import { ChatGoogleGenerativeAI } from "@langchain/google-genai";
import { AgentState, CompetitiveAnalysis } from "../types";
import { safeParseLlmJson, extractTextContent } from "../utils/parseJson";

export const competitiveNode = async (state: AgentState): Promise<Partial<AgentState>> => {
  const companyInfo = state.companyInfo;

  if (!companyInfo?.name) {
    return {
      streamLog: ["❌ Error: No company name available for Competitive Node."],
      competitiveAnalysis: getEmptyAnalysis(),
    };
  }

  const resolvedName = companyInfo.name;
  const industry = companyInfo.industry || "their sector";
  // webAnalysis is guaranteed to be present — competitive runs AFTER webSearch
  const webSummary =
    state.webAnalysis?.sourceSummary ||
    "No web search summary available. Use general knowledge about the company.";
  const recentEvents =
    state.webAnalysis?.recentEvents?.length
      ? state.webAnalysis.recentEvents.join(", ")
      : "No recent events from search.";

  try {
    const model = new ChatGoogleGenerativeAI({
      model: "gemini-2.5-flash",
      temperature: 0.1,
      maxOutputTokens: 2048,
    });

    const systemPrompt =
      "You are a strategy consultant specializing in competitive analysis.\nReturn ONLY valid JSON, no markdown.";
    const userPrompt = `Analyze the competitive position of ${resolvedName} in the ${industry} industry.\n\nContext from web research:\n${webSummary}\nRecent events: ${recentEvents}\n\nReturn this exact JSON:\n{\n  "mainCompetitors": ["string"],\n  "marketPosition": "leader" | "challenger" | "niche" | "emerging",\n  "moatType": "brand" | "network_effects" | "cost_advantage" | "switching_costs" | "IP" | "none",\n  "moatStrength": number,\n  "moatRationale": "string",\n  "differentiators": ["string"],\n  "threats": ["string"],\n  "marketSizeTAM": "string | null",\n  "competitiveScore": number\n}\n\nmoatStrength and competitiveScore must be 1-10.\nBe specific. Use real competitor names.\nBase moat score strictly on evidence from context.`;

    const response = await model.invoke([
      { role: "system", content: systemPrompt },
      { role: "user", content: userPrompt },
    ]);

    const raw = extractTextContent(response.content);
    const { data: parsedResult, error } = safeParseLlmJson(raw, "CompetitiveNode");

    const result = (parsedResult ?? {}) as Record<string, unknown>;
    if (error) {
      console.warn("[CompetitiveNode] Using fallback analysis due to parse error.");
    }

    const analysis: CompetitiveAnalysis = {
      mainCompetitors: Array.isArray(result.mainCompetitors) ? result.mainCompetitors as string[] : [],
      marketPosition:
        (result.marketPosition as CompetitiveAnalysis["marketPosition"]) || "challenger",
      moatType: (result.moatType as CompetitiveAnalysis["moatType"]) || "none",
      moatStrength: typeof result.moatStrength === "number" ? result.moatStrength : 5,
      moatRationale: (result.moatRationale as string) || "No data available.",
      differentiators: Array.isArray(result.differentiators) ? result.differentiators as string[] : [],
      threats: Array.isArray(result.threats) ? result.threats as string[] : [],
      marketSizeTAM: (result.marketSizeTAM as string) || null,
      competitiveScore: typeof result.competitiveScore === "number" ? result.competitiveScore : 5,
    };

    return {
      competitiveAnalysis: analysis,
      streamLog: [
        `✅ Competitive analysis done. \n  Moat: ${analysis.moatType} (${analysis.moatStrength}/10)`,
      ],
    };
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    return {
      streamLog: [`❌ Error in Competitive Node: ${errorMessage}`],
      competitiveAnalysis: getEmptyAnalysis(),
    };
  }
};

function getEmptyAnalysis(): CompetitiveAnalysis {
  return {
    mainCompetitors: [],
    marketPosition: "challenger",
    moatType: "none",
    moatStrength: 5,
    moatRationale: "Unknown due to error.",
    differentiators: [],
    threats: [],
    marketSizeTAM: null,
    competitiveScore: 5,
  };
}

File: backend/nodes/competitive.py
"""
nodes/competitive.py — Competitive Analysis Node
Port of src/lib/nodes/competitive.ts
"""

import logging
from langchain_groq import ChatGroq
from langchain_core.messages import SystemMessage, HumanMessage

from state import AgentState
from utils.parse_json import safe_parse_llm_json

logger = logging.getLogger(__name__)


def _empty_analysis() -> dict:
    return {
        "mainCompetitors": [],
        "marketPosition": "challenger",
        "moatType": "none",
        "moatStrength": 5,
        "moatRationale": "Unknown due to error.",
        "differentiators": [],
        "threats": [],
        "marketSizeTAM": None,
        "competitiveScore": 5,
    }


async def competitive_node(state: AgentState) -> dict:
    company_info = state.get("companyInfo") or {}

    if not company_info.get("name"):
        return {
            "streamLog": ["⊗ ABORT — No entity resolved for Competitive."],
            "competitiveAnalysis": _empty_analysis(),
        }

    resolved_name = company_info["name"]
    industry = company_info.get("industry") or "their sector"

    # competitive runs AFTER webSearch — webAnalysis is guaranteed to be present
    web_analysis = state.get("webAnalysis") or {}
    web_summary = web_analysis.get("sourceSummary") or "No web search summary available. Use general knowledge about the company."
    recent_events_list = web_analysis.get("recentEvents") or []
    recent_events = ", ".join(recent_events_list) if recent_events_list else "No recent events from search."

    try:
        model = ChatGroq(
            model="llama-3.3-70b-versatile",
            temperature=0.1,
            max_tokens=2048,
        )

        system_prompt = (
            "You are a strategy consultant specializing in competitive analysis.\n"
            "Return ONLY valid JSON, no markdown."
        )
        user_prompt = (
            f"Analyze the competitive position of {resolved_name} in the {industry} industry.\n\n"
            f"Context from web research:\n{web_summary}\n"
            f"Recent events: {recent_events}\n\n"
            f"Return this exact JSON:\n"
            f'{{\n'
            f'  "mainCompetitors": ["string"],\n'
            f'  "marketPosition": "leader" | "challenger" | "niche" | "emerging",\n'
            f'  "moatType": "brand" | "network_effects" | "cost_advantage" | "switching_costs" | "IP" | "none",\n'
            f'  "moatStrength": number,\n'
            f'  "moatRationale": "string",\n'
            f'  "differentiators": ["string"],\n'
            f'  "threats": ["string"],\n'
            f'  "marketSizeTAM": "string | null",\n'
            f'  "competitiveScore": number\n'
            f'}}\n\n'
            f"moatStrength and competitiveScore must be 1-10.\n"
            f"Be specific. Use real competitor names.\n"
            f"Base moat score strictly on evidence from context."
        )

        response = await model.ainvoke([
            SystemMessage(content=system_prompt),
            HumanMessage(content=user_prompt),
        ])
        raw = response.content if isinstance(response.content, str) else str(response.content)
        parsed_result, error = safe_parse_llm_json(raw, "CompetitiveNode")

        result = parsed_result or {}
        if error:
            logger.warning("[CompetitiveNode] Using fallback analysis due to parse error.")

        moat_strength = result.get("moatStrength", 5)
        if not isinstance(moat_strength, (int, float)):
            moat_strength = 5
        competitive_score = result.get("competitiveScore", 5)
        if not isinstance(competitive_score, (int, float)):
            competitive_score = 5

        analysis = {
            "mainCompetitors": result.get("mainCompetitors") if isinstance(result.get("mainCompetitors"), list) else [],
            "marketPosition": result.get("marketPosition") or "challenger",
            "moatType": result.get("moatType") or "none",
            "moatStrength": moat_strength,
            "moatRationale": result.get("moatRationale") or "No data available.",
            "differentiators": result.get("differentiators") if isinstance(result.get("differentiators"), list) else [],
            "threats": result.get("threats") if isinstance(result.get("threats"), list) else [],
            "marketSizeTAM": result.get("marketSizeTAM") or None,
            "competitiveScore": competitive_score,
        }

        return {
            "competitiveAnalysis": analysis,
            "streamLog": [f"◈ COMPETITIVE MAPPED. \n  Moat: {analysis['moatType']} ({analysis['moatStrength']}/10)"],
        }

    except Exception as exc:
        error_message = str(exc)
        return {
            "streamLog": [f"⊗ COMPETITIVE FAULT — {error_message}"],
            "competitiveAnalysis": _empty_analysis(),
        }
